Iterate on 'Insert link' interface

noisysocks opened the pull request:
Closes #4572.

![linky-links](https://user-images.githubusercontent.com/612155/38406719-bec60cdc-39b9-11e8-9da1-6bc912d294a7.gif)

- Restores the _Open in new window_ setting by reverting #4583.
- Removes the _Unlink_ button. Instead, links can be removed by toggling off the _Link_ button in the formatting toolbar.
- Styles the _Link Settings_ button to look like the mockup in https://github.com/WordPress/gutenberg/issues/4572.
- Styles the suggested links dropdown to look like the mockup in https://github.com/WordPress/gutenberg/issues/4572.
- Makes it so that the UI expands to fit long URLs when not in editing mode.
- Keeps the _Edit link_ and _Apply_ buttons because I agree with https://github.com/WordPress/gutenberg/issues/4572#issuecomment-360538437.
- Does **not** highlight the selected text while the user is inserting a link. I couldn't make it work—see https://github.com/WordPress/gutenberg/pull/4909#issuecomment-364360350 😢 

#### Testing

1. Create a paragraph block.
2. Highlight some text and click the _Link_ button in the formatting toolbar.
3. Create a link by searching for a page.
4. Create a link by typing in an URL.
5. Toggle on/off _Open in new window_ and check that the generated markup receives a `target="_blank"`.
7. Select an existing link and edit it by clicking on the _Edit_ button in the popup.
6. Select an existing link and remove it by clicking on the _Unlink_ button in the formatting toolbar.

noisysocks committed:
Revert "Remove "Open in new window" link option (#4583)"

This reverts commit 5ab37b76f2aebbc38a5c9d8ba5fa298ddf04e23e.

noisysocks committed:
Fix 'Open in new window' link option

Some changes had occured elsewhere in the codebase in between when this
UI was reverted and when it was unreverted.

- Imports `ToggleControl` correctly
- Fixes `RichText` tests

noisysocks committed:
Use ellipsis icon for opening link settings

Change the icon to be the same icon used elsewhere in the editor, and
move the toggle button to the left of the link UI.

noisysocks committed:
Use link toolbar button to remove links

Consolidate the dedicated 'remove link' button into the toolbar button
that adds a link.

noisysocks committed:
After editing a link, exit editing mode

Reset the link UI once the user has edited the link value.

noisysocks committed:
Fix error when setting 'Open in new window' on a new link

Don't call `onChange()` when 'Open in new window' is toggled if we are
adding a new link (as opposed to editing an existing link).

noisysocks committed:
Expand link UI to fit contents

The link popup should widen to fit the URL within it.

noisysocks committed:
Improve suggested links UI in link inserter

Make the suggested links UI more stylistically consistent with the rest
of the editor.

noisysocks commented on blocks/rich-text/format-toolbar/index.js:
```diff
@@ -124,70 +145,101 @@ class FormatToolbar extends Component {
+					const isActive = isFormatActive || isAddingLink;
+					return {
+						...control,
+						icon: isFormatActive ? 'editor-unlink' : 'admin-links', // TODO: Need proper unlink icon
```
I've asked @karmatosed for this icon 😏

iseulde commented:
> Restores the Open in new window setting by reverting #4583. 

😭 but ok. Wouldn't the ellipsis be better placed on the other side? Similar to blocks, and doesn't deserve first focus imho.

karmatosed commented:
> Wouldn't the ellipsis be better placed on the other side? Similar to blocks, and doesn't deserve first focus imho.

I do like it before but not oppose to having secondary after edit, it does feel more the same as our current patterns so lets do that change @noisysocks.

Can the link icon be only at full opacity once then text is selected?  This could fix some issues with people just having no text selected see: #5960

noisysocks committed:
Move settings button to right of the link UI

Move the ellipsis button to the right of the edit button. This is more
consistent with where the ellipsis icon is used elsewhere in the editor.

noisysocks commented:
>I do like it before but not oppose to having secondary after edit, it does feel more the same as our current patterns so lets do that change @noisysocks.

👌 done in 0bb6c854d92a78ddb3b1d7a52f3a7493f1f5600c.

> Can the link icon be only at full opacity once then text is selected? This could fix some issues with people just having no text selected see: #5960

Let's treat this as a seperate issue since #5960 is open and I have some questions about it (namely https://github.com/WordPress/gutenberg/issues/5960#issuecomment-379604016).

karmatosed reviewed (approved):
Looks great to me, lets get this in and take from there any iterations, I agree.

karmatosed commented:
@noisysocks we seem to have a fun bug :) Dancing link interface:

![2018-04-10 18 02 19](https://user-images.githubusercontent.com/253067/38571795-67e66758-3ce9-11e8-8e63-106ec64fa6e2.gif)

Happens for me in Chrome and Firefox. Props to @mtias for finding this one.
